File: 5/src/main/kotlin/main.kt
import java.io.File
import java.util.*
import kotlin.system.measureTimeMillis

fun main() {
    val ms = measureTimeMillis {
        val input = File("input.txt").readLines().first()
        println("Part 1\t: ${react(input, null)}")
        println("Part 2\t: ${fullyReact(input)}")
    }
    println(ms)
}

fun react(input: String, skip: Char?): Int {
    val stack = Stack<Char>()
    input.toCharArray().forEach {
        if (skip != null && (skip == it || skip.toInt().xor(it.toInt()) == 32)) {
            //Skip
        } else {
            if (!stack.empty() && stack.peek().toInt().xor(it.toInt()) == 32) {
                stack.pop()
            } else {
                stack.push(it)
            }
        }
    }
    return stack.size
}

fun fullyReact(input: String): Int {
    return ('a' until 'z').asSequence().map { react(input, it) }.min()!!
}

File: 5/src/main.rs
use std::fs::File;
use std::io::Read;

fn main() {
    let mut file = File::open("/app/5/input.txt").expect("File not found");
    let mut input = String::new();
    file.read_to_string(&mut input).expect("Unable to read file");
    input = input.trim().to_string();
    let part1 = react(&input, None);
    println!("Part 1: {}", part1.len());
    println!("Part 2: {}", fully_react(&part1));
}

fn react(input: &str, skip: Option<char>) -> String {
  let mut stack = Vec::<char>::with_capacity(2^18);
  for c in input.chars() {
    if skip.is_some() && (skip.unwrap() == c || reacts(c, skip.unwrap())) {
      //skip
    } else {
      if !stack.is_empty() && reacts(*stack.last().unwrap(), c) {
        stack.pop();
      } else {
        stack.push(c);
      }
    }
  }
  return stack.into_iter().collect()
}

fn fully_react(input: &str) -> usize {
  return (b'a'..=b'z').map(|c| react(&input, Some(c as char)).len()).min().unwrap()
}

fn reacts(a: char, b: char) -> bool {
    return a != b && a.eq_ignore_ascii_case(&b);
}

